### ISSUE TITLE
virt-manager: we should stop forking

### ISSUE THREAD

**Issue by crobinso** [virt-manager/virt-manager#731]
virt-manager default behavior is to fork itself, call `os.setsid()`, and point stdin/stdout/stderr fds to /dev/null. This is disabled with `--no-fork` or `--debug`. The app has been like this for well over a decade.

The reason we fork, is to force ssh to _not_ prompt for password on the terminal, if ssh keys aren't set up. This applies to libvirt ssh URIs, and ssh we launch directly to connect to VM graphical console. We can't just close file descriptors, we need to actually disconnect from the controlling terminal to force ssh to not write there. Hence the fork and setsid.

This `fork and continue on` behavior has caused a lot of problems over the years. Most recently with macos https://github.com/virt-manager/virt-manager/issues/620 but there were many linux specific issues back in the past. See the wacky way we do imports in virtmanager.py. I wouldn't be surprised if an eventual gtk4 port brought on more issues.

openssh 8.4p1 released in Sep 2020 finally added a feature to force using SSH_ASKPASS: https://man.openbsd.org/ssh.1#SSH_ASKPASS_REQUIRE  However it doesn't work with libvirt since that env variable isn't passed to ssh. I sent a [patch](https://lists.libvirt.org/archives/list/devel@lists.libvirt.org/thread/FVKSYSLKAFMWRXGK345DDDC2IVW2MMU3/) just now

I'd like to find a way to drop forking entirely. Some unorganized thoughts:

* Yes, users should be using ssh keys, but it's not really about that. We want to avoid the case that virt-manager appears to hang because a password prompt is printed somewhere out of sight. Getting an askpass prompt is an improvement over that situation. But so is ssh attempting to launch askpass, failing to find one, and throwing an error, which virt-manager then scrapes and recommends to set up ssh keys (we already do this).

* That said, I'm not sure what modern scenario the ssh console printing happens in, besides running virt-manager from terminal. I feel like this was required once upon a time for virt-manager run from a gnome session, but that's not the case anymore. Looking at ssh code, the main thing that determines if it attempts askpass is if opening /dev/tty fails or not. I assume launching apps from all desktop environments would _not_ have /dev/tty available. So maybe we just don't need to care about this? But if I'm wrong, and we rip out the forking, it could be a maintenance pain if new users try out the app, try to connect to another machine, and virt-manager just appears to hang. So I'm conflicted.

* SSH_ASKPASS_REQUIRE=force will basically give us everything we want. If the libvirt support went back to 2020 like the ssh support does, I'd rip out forking right now.

* We can probably make this stuff behave better in the interim: fork + re-exec virt-manager to avoid issues with dirty post fork state will probably fix all the issues we ever hit, and let us remove a bunch of virtmanager.py hacks. Or maybe we can skip fork and use TIOCNOTTY which may be enough to trick ssh. But both of those may bring up new issues

* We could also use no_tty=1 for libvirt URIs and `-o BatchMode=yes` for our ssh invocations to avoid the console printing problem, but then that breaks any possibility of using askpass ever, so I'd prefer not to do that.

So I'm not really sure what to do. If we wait a year I'd be fine just using `SSH_ASKPASS_REQUIRE=force` once libvirt support trickles out. But I could be convinced to just rip out forking now if we are pretty sure the only case that's affected is running virt-manager from terminal.

cc @berrange @phrdina

**Comment by crobinso** [comment]
Thinking some more, I think for next release we can do this:

* Set SSH_ASKPASS_REQUIRE=force unconditionally
* Add `--fork` and make `--no-fork` the default
* Don't touch the forking logic at all
* Add an env var like VIRTMANAGER_FORCE_FORK=1 which makes `--fork` the default again, as an escape hatch if for some reason people still need it
* Add a man page section about it all, documenting ssh and libvirt version requirements, telling people to file issues if they need forking
* future release: rip all the forking out

**Comment by berrange** [comment]
A while ago I experimented with porting virt-manager to follow the GTK standard practice for applications, but never got around to submitting it as it was quite messy.

TL;DR: we should make ourselves derive from GtkApplication, and move all command line handling into that. We have various CLI args to triggering UI actions (eg display a specific window). These should all turn into GtkApplication actions which are automagically exposed over DBus.  The windows should also get derived from GtkApplicationWindow, so they get action support too.

GtkApplication can be setup so it is "single instance" such that the first instance remains in the foreground, and subsequent  launches just forward CLI args to the original instance and then exit.

This conversion would put us on the path for further useful cleanups to make better use of what Gtk offers. In doing this though, it would be preferrable to entirely remove the --no-fork and NOT add any back compat --fork option.

**Comment by crobinso** [comment]
@berrange maybe I'm misunderstanding, but virt-manager already does that: we use GtkApplication and run as a single instance. the old manual dbus API stuff was converted to GtkApplication framework too.

**Comment by berrange** [comment]
> @berrange maybe I'm misunderstanding, but virt-manager already does that: we use GtkApplication and run as a single instance. the old manual dbus API stuff was converted to GtkApplication framework too.

Yeah, i didn't explain it sufficiently.  The way we currently use GtkApplication is pretty limited and not really taking advantage of its functionality fully. ie we create an application instance internally from vmmEngine, and then pass over some args in the 'handle_cli' action.

What I mean is making vmmEngine a subclass of GtkApplication, and moving CLI processing into the "command_line" and/or "local_command_line" virtual methods defined by GtkApplication, instead of re-inventing this via the 'handle_cli" action. Then define actions for each distinct action. ie a "show_manager" action, a "show_console(vm)" action, etc.  Dropping our "fork"  logic is then more straightforward.
